### ISSUE TITLE
[Docker] [Autoscaler] Failure to setup head node on recycled head

### ISSUE THREAD

**Issue by amholler** [ray-project/ray#17228]
Thanks to @richardliaw for consultation, any mistakes mine.

### What is the problem?
Failure to setup head node when running "ray up" on recycled head: cannot create directory '/tmp/ray_tmp_mount/default/~'

### Reproduction (REQUIRED)
ray up test-aws.yaml

Details:
Details:
 I had run "ray up test-aws.yaml" (will attach yaml) several times in the past, but today it started failing consistently.
 Adding -vvv we see (abbreviated):
```
 ubuntu@ip-172-31-12-183:~$ ray up -vvv test-aws.yaml
 <snip>
   Running `docker run --rm --name ray_container -d -it -v /tmp/ray_tmp_mount/default//home/ray/.aws:/home/ray/.aws -v /tmp/ray_tmp_mount/default//home/ubuntu/ludwig/:/home/ubuntu/ludwig/ -v /tmp/ray_tmp_mount/default//home/ray/.kaggle:/home/ray/.kaggle -e LC_ALL=C.UTF-8 -e LANG=C.UTF-8 --shm-size='21627607203.84b' --runtime=nvidia --net=host ludwigai/ludwig-ray-gpu:master bash`
     Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@34.221.222.129 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (docker run --rm --name ray_container -d -it -v /tmp/ray_tmp_mount/default//home/ray/.aws:/home/ray/.aws -v /tmp/ray_tmp_mount/default//home/ubuntu/ludwig/:/home/ubuntu/ludwig/ -v /tmp/ray_tmp_mount/default//home/ray/.kaggle:/home/ray/.kaggle -e LC_ALL=C.UTF-8 -e LANG=C.UTF-8 --shm-size='"'"'21627607203.84b'"'"' --runtime=nvidia --net=host ludwigai/ludwig-ray-gpu:master bash)'`
 9984013237ed478e3dd50c96a0137e31dc6564f2e0b1b47d187eb3902b0f0bb4
 Shared connection to 34.221.222.129 closed.
   Running `mkdir -p /tmp/ray_tmp_mount/default/~ && chown -R ubuntu /tmp/ray_tmp_mount/default/~`
     Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@34.221.222.129 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (mkdir -p /tmp/ray_tmp_mount/default/~ && chown -R ubuntu /tmp/ray_tmp_mount/default/~)'`
 mkdir: cannot create directory '/tmp/ray_tmp_mount/default/~': Permission denied
 Shared connection to 34.221.222.129 closed.
   New status: update-failed
   !!!
   {'message': 'SSH command failed.'}
   SSH command failed.
   !!!  Failed to setup head node.
 ubuntu@ip-172-31-12-183:~$

My file mounts are:
 file_mounts:
     {
         /home/ray/.aws: /home/ubuntu/.aws,
         /home/ubuntu/ludwig/: /home/ubuntu/ludwig,
         /home/ray/.kaggle: /home/ubuntu/.kaggle,
     }
Looking at the ray head:
 ubuntu@ip-172-31-16-91:~$ ls -alR /tmp/ray_tmp_mount/
 /tmp/ray_tmp_mount/:
 total 36
 drwxr-xr-x  3 root root  4096 Jul 20 21:33 .
 drwxrwxrwt 10 root root 28672 Jul 20 21:43 ..
 drwxr-xr-x  3 root root  4096 Jul 20 21:33 default/tmp/ray_tmp_mount/default:
 total 12
 drwxr-xr-x 3 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 3 root root 4096 Jul 20 21:33 ..
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 home/tmp/ray_tmp_mount/default/home:
 total 16
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 3 root root 4096 Jul 20 21:33 ..
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 ray
 drwxr-xr-x 3 root root 4096 Jul 20 21:33 ubuntu/tmp/ray_tmp_mount/default/home/ray:
 total 16
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 ..
 drwxr-xr-x 2 root root 4096 Jul 20 21:33 .aws
 drwxr-xr-x 2 root root 4096 Jul 20 21:33 .kaggle/tmp/ray_tmp_mount/default/home/ray/.aws:
 total 8
 drwxr-xr-x 2 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 ../tmp/ray_tmp_mount/default/home/ray/.kaggle:
 total 8
 drwxr-xr-x 2 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 ../tmp/ray_tmp_mount/default/home/ubuntu:
 total 12
 drwxr-xr-x 3 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 4 root root 4096 Jul 20 21:33 ..
 drwxr-xr-x 2 root root 4096 Jul 20 21:33 ludwig/tmp/ray_tmp_mount/default/home/ubuntu/ludwig:
 total 8
 drwxr-xr-x 2 root root 4096 Jul 20 21:33 .
 drwxr-xr-x 3 root root 4096 Jul 20 21:33 ..
```
Deleting the /tmp/ray_tmp_mount directory from the cached VM did not make any difference.

Setting `cache_stopped_nodes: False` under `provider` in the cluster configuration made the problem go away.
In this case, the /tmp/ray_tmp_mount related code is not executed

**Comment by amholler** [comment]
[test-aws.yaml.txt](https://github.com/ray-project/ray/files/6851915/test-aws.yaml.txt)

**Comment by richardliaw** [comment]
@ijrsvt @mwtian can you take a look?

**Comment by ijrsvt** [comment]
Hmm @amholler the `~/` file_mounts are created by Ray for the files to allow the cluster to autoscale. 

Could you print the full logs if you have them? 

> Deleting the /tmp/ray_tmp_mount directory from the cached VM did not make any difference.

When you re-ran it did it use the same VM as before (sometimes it just cycles to other terminated VMs)?

**Comment by amholler** [comment]
Hi, @ijrsvt, here is the info you requested

>When you re-ran it did it use the same VM as before (sometimes it just cycles to other terminated VMs)?
It used the same VM; I only had one ray head node VM as I confirmed via output from "aws ec2 describe-instances"

>Could you print the full logs if you have them?
```
ubuntu@ip-172-31-12-183:~$ ray up -vvv test-aws.yaml 
Cluster: default

Converting legacy cluster config to a multi node type cluster config. Multi-node-type cluster configs are the recommended format for configuring Ray clusters. See the docs for more information:
https://docs.ray.io/en/master/cluster/config.html#full-configuration
Loaded cached provider configuration from /tmp/ray-config-66723b3b89afaa56ba2bc58c04424d839b19c5ba
If you experience issues with the cloud provider, try re-running the command with --no-config-cache.
AWS config
  IAM Profile: ray-autoscaler-v1 [default]
  EC2 Key pair (all available node types): aws-ludwig
  VPC Subnets (all available node types): subnet-b2343999, subnet-80a134dd, subnet-57f74d2f, subnet-494de903 [default]
  EC2 Security groups (all available node types): sg-0be8bbc08e80f059e [default]
  EC2 AMI (all available node types): ami-064696901389beb84

No head node found. Launching a new cluster. Confirm [y/N]: Y

Acquiring an up-to-date head node
  Reusing nodes i-015f90787489b495d. To disable reuse, set `cache_stopped_nodes: False` under `provider` in the cluster configuration.
  Stopping instances to reuse
  Launched a new head node
  Fetching the new head node
  
<1/1> Setting up head node
  Prepared bootstrap config
  New status: waiting-for-ssh
  [1/7] Waiting for SSH to become available
    Running `uptime` as a test.
    Waiting for IP
      Not yet available, retrying in 5 seconds
      Received: 34.217.25.206
    Running `uptime`
      Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=5s ubuntu@34.217.25.206 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (uptime)'`
ssh: connect to host 34.217.25.206 port 22: Connection timed out
    SSH still not available (SSH command failed.), retrying in 5 seconds.
    Running `uptime`
      Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=5s ubuntu@34.217.25.206 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (uptime)'`
ssh: connect to host 34.217.25.206 port 22: Connection refused
    SSH still not available (SSH command failed.), retrying in 5 seconds.
    Running `uptime`
      Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=5s ubuntu@34.217.25.206 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (uptime)'`
Warning: Permanently added '34.217.25.206' (ECDSA) to the list of known hosts.
 21:51:00 up 0 min,  1 user,  load average: 0.15, 0.03, 0.01
Shared connection to 34.217.25.206 closed.
    Success.
  Running `command -v docker || echo 'NoExist'`
    Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@34.217.25.206 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (command -v docker || echo '"'"'NoExist'"'"')'`
Shared connection to 34.217.25.206 closed.
  Running `docker pull ludwigai/ludwig-ray-gpu:master`
    Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@34.217.25.206 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (docker pull ludwigai/ludwig-ray-gpu:master)'`
master: Pulling from ludwigai/ludwig-ray-gpu
Digest: sha256:41596cac12360b74c12ae286a6e4187b9d83725c275b1f60d56e226c5ddaa8b7
Status: Image is up to date for ludwigai/ludwig-ray-gpu:master
docker.io/ludwigai/ludwig-ray-gpu:master
Shared connection to 34.217.25.206 closed.
  Running `docker inspect -f '{{.State.Running}}' ray_container || true`
    Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@34.217.25.206 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (docker inspect -f '"'"'{{.State.Running}}'"'"' ray_container || true)'`
Shared connection to 34.217.25.206 closed.
  Running `docker inspect -f '{{json .Config.Env}}' ludwigai/ludwig-ray-gpu:master`
    Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@34.217.25.206 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (docker inspect -f '"'"'{{json .Config.Env}}'"'"' ludwigai/ludwig-ray-gpu:master)'`
Shared connection to 34.217.25.206 closed.
  Running `cat /proc/meminfo || true`
    Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@34.217.25.206 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (cat /proc/meminfo || true)'`
Shared connection to 34.217.25.206 closed.
  Running `docker info -f '{{.Runtimes}}' `
    Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@34.217.25.206 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (docker info -f '"'"'{{.Runtimes}}'"'"' )'`
Shared connection to 34.217.25.206 closed.
  Running `nvidia-smi`
    Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@34.217.25.206 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (nvidia-smi)'`
Tue Jul 20 21:51:05 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 450.119.03   Driver Version: 450.119.03   CUDA Version: 11.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            On   | 00000000:00:1E.0 Off |                    0 |
| N/A   39C    P0    25W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
Shared connection to 34.217.25.206 closed.
  Running `docker run --rm --name ray_container -d -it -v /tmp/ray_tmp_mount/default//home/ray/.aws:/home/ray/.aws -v /tmp/ray_tmp_mount/default//home/ubuntu/ludwig/:/home/ubuntu/ludwig/ -v /tmp/ray_tmp_mount/default//home/ray/.kaggle:/home/ray/.kaggle -e LC_ALL=C.UTF-8 -e LANG=C.UTF-8 --shm-size='21628537159.68b' --runtime=nvidia --net=host ludwigai/ludwig-ray-gpu:master bash`
    Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@34.217.25.206 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (docker run --rm --name ray_container -d -it -v /tmp/ray_tmp_mount/default//home/ray/.aws:/home/ray/.aws -v /tmp/ray_tmp_mount/default//home/ubuntu/ludwig/:/home/ubuntu/ludwig/ -v /tmp/ray_tmp_mount/default//home/ray/.kaggle:/home/ray/.kaggle -e LC_ALL=C.UTF-8 -e LANG=C.UTF-8 --shm-size='"'"'21628537159.68b'"'"' --runtime=nvidia --net=host ludwigai/ludwig-ray-gpu:master bash)'`
0d7867efadd5a457c0026a6ab7f36c4a7c13d11ee3dcfa8a978ac5d6cd67b879
Shared connection to 34.217.25.206 closed.
  Running `mkdir -p /tmp/ray_tmp_mount/default/~ && chown -R ubuntu /tmp/ray_tmp_mount/default/~`
    Full command is `ssh -tt -i ~/.ssh/aws-ludwig.pem -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o IdentitiesOnly=yes -o ExitOnForwardFailure=yes -o ServerAliveInterval=5 -o ServerAliveCountMax=3 -o ControlMaster=auto -o ControlPath=/tmp/ray_ssh_1d41c853af/c21f969b5f/%C -o ControlPersist=10s -o ConnectTimeout=120s ubuntu@34.217.25.206 bash --login -c -i 'true && source ~/.bashrc && export OMP_NUM_THREADS=1 PYTHONWARNINGS=ignore && (mkdir -p /tmp/ray_tmp_mount/default/~ && chown -R ubuntu /tmp/ray_tmp_mount/default/~)'`
mkdir: cannot create directory ‘/tmp/ray_tmp_mount/default/~’: Permission denied
Shared connection to 34.217.25.206 closed.
  New status: update-failed
  !!!
  {'message': 'SSH command failed.'}
  SSH command failed.
  !!!
  
  Failed to setup head node.
```

**Comment by ijrsvt** [comment]
Hey @amholler Sorry for the delay, but I have a solution for this problem: 
https://github.com/ray-project/ray/pull/17361
